jsonlogic.go

```diff
@@ -519,6 +542,50 @@ func Apply(rule, data io.Reader, result io.Writer) error {
 	return json.NewEncoder(result).Encode(output)
 }
 
+func (e *Engine) Apply(rule, data io.Reader, result io.Writer) error {
+    if data == nil {
+        data = strings.NewReader("{}")
+    }
+
+    var _rule interface{}
+    var _data interface{}
+
+    ruleStr := new(strings.Builder)
+    _, err := io.Copy(ruleStr, rule)
+    if err != nil {
+        return err
+    }
+
+    ruleKey := ruleStr.String()
+
+    e.mu.Lock()
+    if cachedRule, exists := e.rulesCache[ruleKey]; exists {
+        _rule = cachedRule
+    } else {
+        decoder := json.NewDecoder(strings.NewReader(ruleKey))
```

Comment:
Can this be moved to the build method?

Reply:
@beeme1mr The code decodes the rule only if it hasn’t been decoded yet, acting as a fallback. I’m not sure how we could avoid calling the Build function in such cases.

Reply:
I've moved the logic into the build method